### PR TITLE
Add HP Smart Storage Array python plugin

### PR DESCRIPTION
**Pull request by gnoddep**
##### Summary

This module collects controller, logical and physical devices health metrics from HP Smart Storage Arrays.

##### Component Name

hpssa python plugin

##### Additional Information

We run a couple HP servers with Smart Storage Arrays, this enables us to monitor the state and health of the SSAs and disks.

### CODE REVIEW

File: collectors/python.d.plugin/hpssa/hpssa.chart.py [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+                self.error('Cannot locate "{}" binary'.format(self.ssacli_path))
+                return False
+
+        if self.use_sudo:
+            sudo = find_binary('sudo')
+            if not sudo:
+                self.error('Cannot locate "{}" binary'.format('sudo'))
+                return False
+
+            err = self._get_raw_data(command=[sudo, '-n', '-v'], stderr=True)
```
Comment (ilyam8) [inline collectors/python.d.plugin/hpssa/hpssa.chart.py]: i suggest to use `-n -l`, see

> https://github.com/netdata/netdata/issues/7107

File: collectors/python.d.plugin/hpssa/hpssa.chart.py [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+            'status_complete': None,
+        }
+
+        for line in self:
+            if _mirror_group_regex.match(line):
+                self._parse_ignored_section()
+            elif _adapter_regex.match(line) \
+                    or _array_regex.match(line) \
+                    or _drive_regex.match(line) \
+                    or _ignored_sections_regex.match(line):
```
Comment (ilyam8) [inline collectors/python.d.plugin/hpssa/hpssa.chart.py]: There is `any` built-in function, i suggest to use it

File: collectors/python.d.plugin/hpssa/hpssa.chart.py [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+                    if key == 'Status':
+                        status_match = _status_complete_regex.match(value)
+                        if status_match:
+                            ld['status'] = status_match.group('status') == 'OK'
+                            ld['status_complete'] = float(status_match.group('percentage')) / 100
+                        else:
+                            ld['status'] = value == 'OK'
+                else:
+                    self._rewind()
+                    break
```
Comment (ilyam8) [inline collectors/python.d.plugin/hpssa/hpssa.chart.py]: `parse_logical_drive` method looks complicated to me, a lot of nested if/else. It would be nice if you could make it simpler

**Comment by ilyam8** [comment]
@gnoddep thanks for the contribution!

Could you share the `ssacli ctrl all show config detail` output?

File: collectors/python.d.plugin/hpssa/hpssa.chart.py (hunk 1 above)
Comment (gnoddep) [reply collectors/python.d.plugin/hpssa/hpssa.chart.py]: Even better than the solution in #7107 is that `sudo -l` accepts a command and checks the user is allowed to run that:

```
                   If a command is specified and is permitted by the security
                   policy, the fully-qualified path to the command is
                   displayed along with any command line arguments.  If
                   command is specified but not allowed, sudo will exit with a
                   status value of 1.
```

**Comment by gnoddep** [comment]
> Could you share the `ssacli ctrl all show config detail` output?

[ssacli-P400-1.txt](https://github.com/netdata/netdata/files/3765236/ssacli-P400-1.txt)
[ssacli-P400-2.txt](https://github.com/netdata/netdata/files/3765237/ssacli-P400-2.txt)
[ssacli-P400i-unassigned-drives.txt](https://github.com/netdata/netdata/files/3765238/ssacli-P400i-unassigned-drives.txt)
[ssacli-P410i-1.txt](https://github.com/netdata/netdata/files/3765239/ssacli-P410i-1.txt)
[ssacli-P410i-2.txt](https://github.com/netdata/netdata/files/3765240/ssacli-P410i-2.txt)
[ssacli-P440ar.txt](https://github.com/netdata/netdata/files/3765241/ssacli-P440ar.txt)

File: collectors/python.d.plugin/hpssa/hpssa.conf [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+#     name: myname            # the JOB's name as it will appear at the
+#                             # dashboard (by default is the job_name)
+#                             # JOBs sharing a name are mutually exclusive
+#     update_every: 1         # the JOB's data collection frequency
+#     priority: 60000         # the JOB's order on the dashboard
+#     penalty: yes            # the JOB's penalty
+#     autodetection_retry: 0  # the JOB's re-check interval in seconds
+# ----------------------------------------------------------------------
+# update_every: 5
+# ssacli_path: /usr/sbin/ssacli
```
Comment (Ferroin) [inline collectors/python.d.plugin/hpssa/hpssa.conf]: Please also document this in the comments above.

See https://github.com/netdata/netdata/blob/master/collectors/python.d.plugin/unbound/unbound.conf#L62-L71 for an example of what it should look like.

**Comment by ilyam8** [comment]
@gnoddep overall lgtm but parsing process looks complicated to me, it is hard to follow. I will see if it can be simplified. It will take some time to me.

File: collectors/python.d.plugin/hpssa/README.md [hunk 5]
```diff
@@ -0,0 +1,44 @@
+# hpssa
+
+Module collects controller, logical and physical devices health metrics from HP Smart Storage Arrays.
```
Comment (joelhans) [inline collectors/python.d.plugin/hpssa/README.md]:
```suggestion
This module collects controller, logical, and physical device health metrics from HP Smart Storage Arrays.
```

File: collectors/python.d.plugin/hpssa/README.md [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+## Requirements:
+
+- `ssacli` program
+- `sudo` program
+- `netdata` user needs to be able to sudo the `ssacli` program without password
+
+To grab stats it executes: `sudo -n ssacli ctrl all show config detail`
```
Comment (joelhans) [inline collectors/python.d.plugin/hpssa/README.md]:
```suggestion
To collect metrics, the module executes: `sudo -n ssacli ctrl all show config detail`
```

File: collectors/python.d.plugin/hpssa/README.md [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+It produces:
```
Comment (joelhans) [inline collectors/python.d.plugin/hpssa/README.md]:
```suggestion
This module produces:
```

File: collectors/python.d.plugin/hpssa/README.md [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+1. Controller state and temperature
+2. Cache module state and temperature
+3. Logical drive state
+4. Physical drive state and temperature
```
Comment (joelhans) [inline collectors/python.d.plugin/hpssa/README.md]: I know it's a little strange, but our linter wants two extra spaces between the period and the start of the text.

So:

```
1.   Controller state and temperature
2.   Cache module state and temperature
3.   Logical drive state
4.   Physical drive state and temperature
```

File: collectors/python.d.plugin/hpssa/README.md [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+## Prerequisite
+
+This module uses `ssacli` which can only be executed by root.  It uses
```
Comment (joelhans) [inline collectors/python.d.plugin/hpssa/README.md]:
```suggestion
This module uses `ssacli`, which can only be executed by root. It uses
```

File: collectors/python.d.plugin/hpssa/README.md [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+## Configuration
+
+**hpssa** is disabled by default. Should be explicitly enabled in `python.d.conf`.
+
+```yaml
+hpssa: yes
+```
+
+If `ssacli` cannot be found in the `PATH`, configure it in `hpssa.conf`
```
Comment (joelhans) [inline collectors/python.d.plugin/hpssa/README.md]:
```suggestion
If `ssacli` cannot be found in the `PATH`, configure it in `hpssa.conf`.
```

**Comment by vobruba-martin** [comment]
Could you make the command configurable like it is in python.d/postfix plugin? The `ssacli` utility is called `hpssacli` at least on my Ubuntu servers.

**Comment by gnoddep** [comment]
> Could you make the command configurable like it is in python.d/postfix plugin? The `ssacli` utility is called `hpssacli` at least on my Ubuntu servers.

You can use `ssacli_path` in the configuration file for this

**Comment by ilyam8** [comment]
> Sure, if you can make it better or nicer, that would be great!

Ok, it is not easy, `ssacli ctrl all show config detail` output is not very friendly to parse taking into account all variants from https://github.com/netdata/netdata/pull/7181#issuecomment-545700797. I think we can imrove it later 😄

**Comment by ilyam8** [comment]
@gnoddep i suggest to add some alarms. This module monitors state of controller/ld/pd, i think if state != `OK` is good `warn` condition.

btw what other statuses can there be?

**Comment by ilyam8** [comment]
@gnoddep do you have time [to finish it?](https://github.com/netdata/netdata/pull/7181#issuecomment-552526500)

**Comment by gnoddep** [comment]
> @gnoddep do you have time [to finish it?](https://github.com/netdata/netdata/pull/7181#issuecomment-552526500)

Yes, I will. At the moment I'm quite busy, but I will have some time for this in the next couple of weeks.

**Comment by ilyam8** [comment]
> Yes, I will. At the moment I'm quite busy, but I will have some time for this in the next couple of weeks.

I think this can be done in a separate PR

**Comment by ilyam8** [comment]
@gnoddep thank you!

**Comment by bobatsar** [comment]
I just tried to test the plugin but it doesn't find any controllers:

```
2019-12-18 12:14:47: python.d INFO: hpssa[hpssa] : Command: ['/usr/bin/sudo', '-n', '/usr/sbin/ssacli', 'ctrl', 'all', 'show', 'config', 'detail']
2019-12-18 12:14:51: python.d ERROR: hpssa[hpssa] : Cannot parse line: Smart Array P410i in Slot 0 (Embedded)
2019-12-18 12:14:51: python.d INFO: hpssa[hpssa] : Discovered adapters: []
2019-12-18 12:14:51: python.d ERROR: hpssa[hpssa] : No adapters discovered
2019-12-18 12:14:51: python.d INFO: plugin[main] : hpssa[hpssa] : check failed
```
I have two controllers but none of them seems to be detected:

```
Smart Storage Administrator CLI 3.30.13.0

Smart Array P212 in Slot 5                (sn: PACCPID11170BVJ)
Smart Array P410i in Slot 0 (Embedded)    (sn: 50123456789ABCDE)
```
